### DISCUSSION TITLE
emailRedirectTo not working on auth.updateUser

### DISCUSSION THREAD

**Discussion by iprefernotto** [supabase/supabase#18810] [Questions]
Can't get this redirection working.
  
  Verification mail is sent correctly. But the confirmation links end with 
  
"... 98&type=email_change&redirect_to=http://localhost:3000/"

```
  const { data, error } = await supabase.auth.updateUser({
    email: tempEmailInput.value,
    options: {
      emailRedirectTo: 'http://localhost:3000/success-email-change'
    }
  });
```

I have set the redirection URLs in the Supabase Dashboard:

![sb_authurl](https://github.com/supabase/supabase/assets/84004656/d0e20ffd-2d4b-403a-af20-caec66c91d4c)

Any idea what to do?

**Comment by iprefernotto** [comment]
Would be truly thankful for any.

As far as I can see, I've implemented it exactly like it is explained in the documentation:

https://supabase.com/docs/reference/javascript/auth-updateuser

**Comment by Har1s-Akbar** [comment]
There are 2-3 scenarios here:

1. If you are trying to sign up a user, then you are not using the right method for it.

2. If you are trying to update an already existing user, the **updateUser** method does not support sending the verification email.
> (I have actually tried it and the **updateUser** method does not send verification emails)

3. For the redirect, I tried to recreate your code:
` const { data, error } = await clientSupabase.auth.updateUser({
            email: ${email},
            options: {
              emailRedirectTo: 'http://localhost:3000/any-other-Link'
            }
          });
`
The Vs code is throwing this error:
`Argument of type '{ email: string; options: { emailRedirectTo: string; }; }' is not assignable to parameter of type 'UserAttributes'.
Object literal may only specify known properties, and 'options' does not exist in type 'UserAttributes'.`

Even tho the documents mention the **options** & **emailRedirectTo** but i don't think **auth.updateUser** actually support the redirect email.

(If you can share what you are actually trying to do here, I might be able to help you :) )

**Comment by iprefernotto** [comment]
What I try to accomplish is to **update the email for an authenticated user**. Think of some user wants to update the email of his account.

The documentation tells me I should use:

`const { data, error } = await supabase.auth.updateUser({email: 'new@email.com'})
`
The tabs above the code say exactly '_Update the email for an authenticated user_'.

It does indeed change the email of the user and also send the confirmation mail. But i cannot change the url where the user gets redirected after clicking the confirmation link in the mail. 

Is there any workaround? Thank you for your help. Much appreciated Har1s.

**Comment by rudokotrla** [comment]
Same problem. Call to `const { error } = await supabase.auth.updateUser(attributes, {
    emailRedirectTo: "/guest/password-create"
  })` fires a PUT request:
  ```
{
  "headers": [
    {
      "cf_connecting_ip": "192.168.65.1",
      "user_agent": "node"
    }
  ],
  "method": "PUT",
  "path": "/auth/v1/user?redirect_to=%2Fguest%2Fenter-password",
  "protocol": "HTTP/1.1"
}
```
and the `emailRedirectTo` param seems to be there but the generated link to confirm the change in the email has this format:
`http://127.0.0.1:3000/auth/confirm?token_hash=<nice hash>&type=email_change&next=/login`. I would expect the `next` query param to be same as my specified value in the `emailRedirectTo` param but it seems to be.

**Comment by nassmim** [comment]
I have the same issue. I had my url configured in the "URL configuration" section, and I tried both emailRedirectTo and redirectTo within the updateUser. None worked. 
I ended up simply setting the url directly on the email template, which for my usecase was enough.

**Reply by pat-RR-ick** [reply]
can i please ask how you did this. can you paste your example, i cannot figure this out to update the email template for confirming new email address.

**Reply by nassmim** [reply]
Sorry for my late reply. 
Supabase give you the names of the variables in the section "Authentication/Email template":
{{ .ConfirmationURL }} : URL to confirm the email change
{{ .Token }} : The 6-digit numeric email OTP
{{ .TokenHash }} : The hashed token used in the URL
{{ .SiteURL }} : The URL of the site
{{ .Email }} : The original user's email address
{{ .NewEmail }} : The user's new email address
{{ .Data }} : The user's user_metadata
{{ .RedirectTo }} : The URL of emailRedirectTo passed in options

So you can use them to write the url within the href. From my side, I used {{ .SiteURL }} /<my_api_endpoint>
If the dynamic redirection was working, you could just pass there {{ .RedirectTo }}. 
Since it's not working, I can only use the SiteURL variable, which is linked to the one defined in "URL Configuration". However, since i'm working both locally and on vercel with my teammates, it's not the appropriate solution.

**Comment by MrVibe** [comment]
I had the same issue, for me the workaround was to track onAuthSTateChanged Event and redirect the user to the correct url.

**Comment by pat-RR-ick** [comment]
I found one solution. Just update your main callback url to your main url here: https://supabase.com/dashboard/project/axydrphznnwlrbeawkea/auth/url-configuration

like this 
<img width="926" alt="image" src="https://github.com/user-attachments/assets/79924fe2-7b36-4321-a7bb-54af9bb72738">

**Reply by pat-RR-ick** [reply]
Then you handle callbacks based on what the URL encoded message is from subpabase.

**Comment by nassmim** [comment]
any update on this, stil no answer?

**Comment by dahjson** [comment]
Related issue: https://github.com/supabase/supabase-js/issues/1262

**Comment by jrpersico** [comment]
This is worked for me 👇

1. Set localhost site URL and localhost wildcard for redirect URLs in supabase/config.toml

```
[auth]
site_url = "http://localhost:3000"
additional_redirect_urls = ["http://localhost:3000/**"]
```
2. Use auth.updateUser to change the user's email.

```
const {data, error} = await supabase.auth.updateUser(
  {
    email: "test@test.com",
  },
  {
    emailRedirectTo: "http://localhost:3000/settings",
  }
);
```

**Reply by CallumBoase** [reply]
This helped me too - the structure of the function arguments is different to most other auth methods: passing 2 objects as individual arguments (the second being the options object), rather than passing one object with a nested options object
